obspy/core/util/misc.py

```diff
@@ -639,6 +638,111 @@ def buffered_load_entry_point(dist, group, name):
+        Only return objects stored as attr_name, if None return all.
+    :param has_attr:
+        Only return objects that have attribute has_attr, if None return all.
+
+    .. rubric:: General Usage
+
+    Get a list of all resource_ids contained in an event, the objects they
+    are attached to, and the attribute name on the parent object.
+
+    >>> import obspy
+    >>> from obspy.core.event import ResourceIdentifier
+    >>> cat = obspy.read_events()
+    >>> resource_tuple = list(yield_obj_parent_attr(cat, ResourceIdentifier))
+    """
+
+    ids = set()  # id cache to avoid circular references
+
+    def func(obj, attr=None, parent=None):
+        id_tuple = (id(obj), id(parent))
+
+        # If object/parent combo have not been yielded continue.
+        if id_tuple not in ids:
+            ids.add(id_tuple)
+            # Check if this object is stored as the desired attribute.
+            is_attribute = is_attr is None or attr == is_attr
+            # Check if the object has the desired attribute.
+            has_attribute = has_attr is None or hasattr(obj, has_attr)
+            # Check if isinstance of desired class.
+            is_instance = cls is None or isinstance(obj, cls)
+            # Yield object, parent, and attr if desired conditions are met.
+            if is_attribute and has_attribute and is_instance:
+                yield (obj, parent, attr)
+            # Iterate through basic built-in types.
+            if isinstance(obj, (list, tuple)):
+                for val in obj:
+                    for out in func(val, attr=attr, parent=obj):
+                        yield out
+            elif isinstance(obj, dict):
+                for item, val in obj.items():
+                    for out in func(val, attr=item, parent=obj):
+                        yield out
+            # Iterate through non built-in object attributes.
+            elif hasattr(obj, '__slots__'):
+                for attr in obj.__slots__:
+                    val = getattr(obj, attr)
+                    for out in func(val, attr=attr, parent=obj):
+                        yield out
+            elif hasattr(obj, '__dict__'):
+                for item, val in obj.__dict__.items():
+                    for out in func(val, attr=item, parent=obj):
+                        yield out
+
+    return func(obj)
+
+
+def _yield_resource_id_parent_attr(obj):
+    """
+    Specialized form of yield_obj_parent_attr for getting ResourceIdentifiers.
+
+    This function makes some assumptions because only resource_identifiers are
+    being sought in order to improve efficiency.
+    """
+    from obspy.core.event import ResourceIdentifier
+
+    ids = set()  # id cache to avoid circular references
+
+    def func(obj, attr=None, parent=None):
+
+        if obj is not None and (hasattr(obj, '__dict__') or
+                                isinstance(obj, (list, tuple))):
+            id_tuple = (id(obj), id(parent))
+            if id_tuple not in ids:
```

Comment:
Good idea but this is a bit tricky; this line is not in the scope of a loop so the `continue` keyword isn't valid. Since this is a generator we can just return (which is the new way of raising StopIteration according to PEP 479). I think this will work in 2.7, let's see what the CI says.

Reply:
Ah yea sorry you are right. CI seems to like in on 2.7 (which I think works only when one does not return a value).